squidpy/pl/_spatial_utils.py

```diff
@@ -0,0 +1,960 @@
+
+
+# named tuples
+class FigParams(NamedTuple):
+    """Figure params."""
+
+    fig: Figure
+    ax: Axes
+    axs: Sequence[Axes] | None
+    iter_panels: Tuple[Any, Any]
+    title: _SeqStr
+    ax_labels: Sequence[str]
+    frameon: bool | None
+
+
+class CmapParams(NamedTuple):
+    """Cmap params."""
+
+    cmap: Colormap
+    img_cmap: Colormap
+    norm: Normalize
+
+
+class OutlineParams(NamedTuple):
+    """Outline params."""
+
+    outline: bool
+    gap_size: float
+    gap_color: str
+    bg_size: float
+    bg_color: str
+
+
+class ScalebarParams(NamedTuple):
+    """Scalebar params."""
+
+    scalebar_dx: Sequence[float] | None
+    scalebar_units: _SeqStr
+
+
+class ColorParams(NamedTuple):
+    """Color params."""
+
+    shape: _AvailShapes | None
+    color: Sequence[str | None]
+    groups: Sequence[str] | None
+    alpha: float
+    img_alpha: float
+    use_raw: bool
+
+
+class SpatialParams(NamedTuple):
+    """Color params."""
+
+    library_id: Sequence[str]
+    scale_factor: Sequence[float]
+    size: Sequence[float]
+    img: Sequence[NDArrayA] | Tuple[None, ...]
+    segment: Sequence[NDArrayA] | Tuple[None, ...]
+    cell_id: Sequence[NDArrayA] | Tuple[None, ...]
+
+
+def _get_library_id(
+    adata: AnnData,
+    shape: _AvailShapes | None,
+    spatial_key: str = Key.uns.spatial,
+    library_id: Sequence[str] | None = None,
+    library_key: str | None = None,
+) -> Sequence[str]:
+    if shape is not None:
+        library_id = Key.uns.library_id(adata, spatial_key, library_id, return_all=True)
+        if library_id is None:
+            raise ValueError(f"Could not fetch `library_id`, check that `spatial_key: {spatial_key}` is correct.")
+        return library_id
+    if library_key is not None:
+        if library_key not in adata.obs.columns:
+            raise ValueError(f"`library_key: {library_key}` not in `adata.obs`.")
+        if library_id is None:
+            library_id = adata.obs[library_key].cat.categories.tolist()
+        _assert_value_in_obs(adata, key=library_key, val=library_id)
+        if isinstance(library_id, str):
+            library_id = [library_id]
+        return library_id
+    if library_id is None:
+        logg.warning(
+            "Please specify a valid `library_id` or set it permanently in `adata.uns['spatial'][<library_id>]`"
+        )
+        library_id = [""]
+    elif isinstance(library_id, list):  # get library_id from arg
+        library_id = library_id
+    elif isinstance(library_id, str):
+        library_id = [library_id]
+    else:
+        raise ValueError(f"Invalid `library_id`: {library_id}.")
+    return library_id
+
+
+def _get_image(
+    adata: AnnData,
+    library_id: Sequence[str],
+    spatial_key: str = Key.obsm.spatial,
+    img: _SeqArray | bool = None,
+    img_res_key: str | None = None,
+    img_channel: int | None = None,
+    img_cmap: Colormap | str | None = None,
+) -> Union[Sequence[NDArrayA], Tuple[None, ...]]:
```

Comment:
List[None]

Reply:
mypy throws error cause it's a generator